homeassistant/components/mysensors/gateway.py

```diff
@@ -217,22 +217,26 @@ async def _discover_persistent_devices(
 async def gw_stop(hass, gateway: BaseAsyncGateway):
     """Stop the gateway."""
     _LOGGER.info("stopping gateway %s", gateway.unique_id)
-    if (
-        hasattr(gateway, "connect_task")
-        and gateway.connect_task is not None
-        and not gateway.connect_task.done()
-    ):
-        gateway.connect_task.cancel()
+    connect_task = hass.data.get(
+        MYSENSORS_GATEWAY_START_TASK.format(gateway.unique_id), None
+    )
+    if connect_task is not None and not connect_task.done():
+        connect_task.cancel()
     await gateway.stop()
 
 
 async def _gw_start(hass: HomeAssistantType, gateway: BaseAsyncGateway):
     """Start the gateway."""
     # Don't use hass.async_create_task to avoid holding up setup indefinitely.
-    gateway.connect_task = hass.loop.create_task(gateway.start())
+    hass.data[
+        MYSENSORS_GATEWAY_START_TASK.format(gateway.unique_id)
+    ] = asyncio.create_task(
+        gateway.start()
+    )  # store the connect task so it can be cancelled in gw_stop
 
     hass.bus.async_listen_once(
-        EVENT_HOMEASSISTANT_STOP, lambda event: gw_stop(hass, gateway)
+        EVENT_HOMEASSISTANT_STOP,
+        lambda event: asyncio.create_task(gw_stop(hass, gateway)),
```

Comment:
I think we can have a coroutine function as event callback, which awaits the gateway stop function.